In [ ]:
import pandas as pd
import random
import math
import numpy as np
from scipy.stats import spearmanr
import argparse as ap
from sentence_transformers import SentenceTransformer
import spacy
nlp = spacy.load('en_core_web_sm')

from google.colab import drive
drive.mount('/content/drive')

path = '/content/drive/MyDrive/GIL/Mushoom/'#C:\Users\FLopezP\Documents\GitHub\Mu-SHROOM-GIL\Datasets\GPT_Answers\

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
gpt_answers = pd.read_csv(f'{path}Datasets/QAC_Val.csv')
gpt_answers = gpt_answers.drop(columns = ["Unnamed: 0"])
gpt_answers.head(6)

Pregunta  \
0  In which city were the 26th biathlon world cha...   
1                Who did Dave Franco play in Scrubs?   
2  Which municipalities does the Italian commune ...   
3  What is the UN Sustainable Development Goal 11...   
4  Is the Arts and Humanities Citation Index stil...   
5             What is the home stadium of FC Zhenis?   

                                       Respuesta GPT  \
0  The 26th Biathlon World Championships were hel...   
1  Dave Franco played the character Cole in the f...   
2  Ponzone borders the following municipalities: ...   
3  The provided text does not contain information...   
4  Yes, the Arts and Humanities Citation Index (A...   
5  The home stadium of FC Zhenis is the K.Munaitp...   

                                            Contexto  
0  The 26th Biathlon World Championships were hel...  
1  David John Franco (born June 12, 1985) is an A...  
2  The province of Alessandria (Italian: provinci...  
3  SDG 4, or Sustainable Development Goal 4, is a...  
4  The Arts and Humanities Citation Index (AHCI),...  
5  FC Zhenis (Kazakh: Жеңіс футбол клубы) is a Ka...

In [ ]:
val_ds = pd.read_json(f'{path}Datasets/mushroom.en-val.v2.jsonl', lines = True)
val_ds.head(6)

id lang                                        model_input  \
0  val-en-1   EN  What did Petra van Staveren win a gold medal for?   
1  val-en-2   EN  How many genera does the Erysiphales order con...   
2  val-en-3   EN                   Do all arthropods have antennae?   
3  val-en-4   EN                  When did Chance the Rapper debut?   
4  val-en-5   EN  What is the UN Sustainable Development Goal 11...   
5  val-en-6   EN      What are the four styles of Zhejiang cuisine?   

                                   model_output_text  \
0  Petra van Stoveren won a silver medal in the 2...   
1            The Elysiphale order contains 5 genera.   
2  Yes, all arachnids have antennas. However, not...   
3                 Chance the rapper debuted in 2011.   
4  The UN's Sustainable City initiative defines a...   
5  Zhejing cuisine is known for its unique flavor...   

                    model_id  \
0  tiiuae/falcon-7b-instruct   
1  tiiuae/falcon-7b-instruct   
2  tiiuae/falcon-7b-instruct   
3  tiiuae/falcon-7b-instruct   
4  tiiuae/falcon-7b-instruct   
5  tiiuae/falcon-7b-instruct   

                                         soft_labels  \
0  [{'start': 10, 'prob': 0.2, 'end': 12}, {'star...   
1  [{'start': 4, 'prob': 0.2, 'end': 14}, {'start...   
2  [{'start': 0, 'prob': 0.6000000000000001, 'end...   
3  [{'start': 18, 'prob': 0.0909090909, 'end': 29...   
4  [{'start': 0, 'prob': 0.0909090909, 'end': 9},...   
5  [{'start': 0, 'prob': 0.2727272727, 'end': 7},...   

                                         hard_labels  \
0                     [[25, 31], [45, 49], [69, 83]]   
1                                         [[30, 31]]   
2              [[0, 3], [9, 18], [63, 70], [78, 87]]   
3                                         [[29, 33]]   
4                                        [[70, 227]]   
5  [[101, 109], [154, 158], [163, 170], [210, 220...   

                                 model_output_logits  \
0  [-5.5669536591, -11.90533638, -13.0743436813, ...   
1  [-6.199614048, -13.7564926147, -14.0058326721,...   
2  [-4.8190689087, -16.5279369354, -10.1344690323...   
3  [-6.2555613518, -10.9392805099, -13.1335763931...   
4  [-3.9737114906, -12.9197320938, -7.2846975327,...   
5  [-5.0601181984, -13.6542539597, -13.8530063629...   

                                 model_output_tokens  
0  [Pet, ra, Ġvan, ĠSto, ve, ren, Ġwon, Ġa, Ġsilv...  
1  [The, ĠE, lys, iph, ale, Ġorder, Ġcontains, Ġ,...  
2  [Yes, ,, Ġall, Ġar, ach, n, ids, Ġhave, Ġanten...  
3  [Ch, ance, Ġthe, Ġrapper, Ġdebuted, Ġin, Ġ, 20...  
4  [The, ĠUN, ', s, ĠSustainable, ĠCity, Ġinitiat...  
5  [Z, he, j, ing, Ġcuisine, Ġis, Ġknown, Ġfor, Ġ...

In [ ]:
merge_data = pd.merge(val_ds, gpt_answers, left_on="model_input", right_on="Pregunta")
merge_data = merge_data[['id','model_input', 'model_output_text', 'Respuesta GPT']]
merge_data.head(6)

id                                        model_input  \
0  val-en-1  What did Petra van Staveren win a gold medal for?   
1  val-en-2  How many genera does the Erysiphales order con...   
2  val-en-3                   Do all arthropods have antennae?   
3  val-en-4                  When did Chance the Rapper debut?   
4  val-en-5  What is the UN Sustainable Development Goal 11...   
5  val-en-6      What are the four styles of Zhejiang cuisine?   

                                   model_output_text  \
0  Petra van Stoveren won a silver medal in the 2...   
1            The Elysiphale order contains 5 genera.   
2  Yes, all arachnids have antennas. However, not...   
3                 Chance the rapper debuted in 2011.   
4  The UN's Sustainable City initiative defines a...   
5  Zhejing cuisine is known for its unique flavor...   

                                       Respuesta GPT  
0  Petra van Staveren won a gold medal in the 100...  
1  The Erysiphales order contains one family, Ery...  
2  Not all arthropods have antennae. While most a...  
3  Chance the Rapper debuted with his mixtape "10...  
4  The provided text does not contain information...  
5  The four styles of Zhejiang cuisine are:\n\n1....

In [ ]:
merge_data.iloc[0]['Respuesta GPT'], merge_data.iloc[0]['model_output_text']

('Petra van Staveren won a gold medal in the 100 meter breaststroke at the 1984 Summer Olympics in Los Angeles.',
 'Petra van Stoveren won a silver medal in the 2008 Summer Olympics in Beijing, China.')

In [ ]:
merge_data.iloc[1]['Respuesta GPT'], merge_data.iloc[1]['model_output_text']

('The Erysiphales order contains one family, Erysiphaceae. It includes multiple genera, but the exact number is not specified in the provided text.',
 'The Elysiphale order contains 5 genera.')

# Get labels

Quitar puntos y comas y revisar que si los corchetes tienen secuencias seguidas juntarlos, revisar la lista y verificar que ese cachito no se encuentre como una subpalabra

In [ ]:
def get_ner(text):
  """It returns the entities from a text"""
  doc = nlp(text)
  val = ["NOUN","PROPN","NUM","ADJ",'0','1','2','3','4','5','6','7','8','9', 'Yes', 'Not']
  entities = []
  for token in doc:
      if token.pos_ in val or token.text in val:
        entities.append(token.text)
  return entities

def join_label(label):
  """It joins the labels that the difference is less than 2"""
  label = sorted(label, key=lambda x: x[0])
  final = []
  for i in range(len(label)-1):
    if label[i+1][0] - label[i][1] <= 2:
      final.append([label[i][0],label[i+1][1]])
    else:
      final.append(label[i])
  if final and final[-1][1] < label[-1][1]:
    final.append(label[-1])
  return final


def get_hard_labels(claims, evidences):
  """It returns the index of words that differs in claims and evidences"""
  hard_labels = []
  for c, e in zip(claims, evidences):
    diff = list(set(get_ner(c)) - set(get_ner(e)))
    label = []
    for i in diff:
      start = c.index(i)
      label.append([start, start+len(i)])
    label = join_label(label)
    hard_labels.append(label)
  return hard_labels


In [ ]:
get_hard_labels([merge_data.iloc[0]['model_output_text']],[merge_data.iloc[0]['Respuesta GPT']])

[[[45, 49], [25, 31], [69, 76], [10, 18], [78, 83]]]

In [ ]:
get_hard_labels([merge_data.iloc[1]['model_output_text']],[merge_data.iloc[1]['Respuesta GPT']])

[[[30, 31], [4, 14]]]

In [ ]:
merge_data.iloc[2]['Respuesta GPT'], merge_data.iloc[2]['model_output_text']

('Not all arthropods have antennae. While most arthropods, like insects and crustaceans, possess antennae, some groups, such as certain arachnids (e.g., spiders), do not have them.',
 'Yes, all arachnids have antennas. However, not all of them are visible to the naked eye.')

In [ ]:
get_hard_labels([merge_data.iloc[2]['model_output_text']],[merge_data.iloc[2]['Respuesta GPT']])

[[[84, 87], [78, 83], [0, 3], [24, 32], [63, 70]]]

In [ ]:
merge_data['hard_labels'] = get_hard_labels(merge_data['model_output_text'], merge_data['Respuesta GPT'])
merge_data.head(6)

id                                        model_input  \
0  val-en-1  What did Petra van Staveren win a gold medal for?   
1  val-en-2  How many genera does the Erysiphales order con...   
2  val-en-3                   Do all arthropods have antennae?   
3  val-en-4                  When did Chance the Rapper debut?   
4  val-en-5  What is the UN Sustainable Development Goal 11...   
5  val-en-6      What are the four styles of Zhejiang cuisine?   

                                   model_output_text  \
0  Petra van Stoveren won a silver medal in the 2...   
1            The Elysiphale order contains 5 genera.   
2  Yes, all arachnids have antennas. However, not...   
3                 Chance the rapper debuted in 2011.   
4  The UN's Sustainable City initiative defines a...   
5  Zhejing cuisine is known for its unique flavor...   

                                       Respuesta GPT  \
0  Petra van Staveren won a gold medal in the 100...   
1  The Erysiphales order contains one family, Ery...   
2  Not all arthropods have antennae. While most a...   
3  Chance the Rapper debuted with his mixtape "10...   
4  The provided text does not contain information...   
5  The four styles of Zhejiang cuisine are:\n\n1....   

                                         hard_labels  \
0           [[10, 18], [25, 31], [45, 49], [69, 83]]   
1                                [[4, 14], [30, 31]]   
2             [[0, 3], [24, 32], [63, 70], [78, 87]]   
3                               [[11, 17], [29, 33]]   
4  [[21, 36], [26, 36], [55, 58], [84, 98], [103,...   
5  [[33, 39], [52, 70], [60, 70], [81, 91], [86, ...   

                                         soft_labels  
0  [{'start': 45, 'prob': 0.9999999, 'end': 49}, ...  
1  [{'start': 30, 'prob': 0.333333, 'end': 31}, {...  
2  [{'start': 84, 'prob': 0.333333, 'end': 87}, {...  
3  [{'start': 11, 'prob': 0.9999999, 'end': 17}, ...  
4  [{'start': 169, 'prob': 0.333333, 'end': 174},...  
5  [{'start': 154, 'prob': 0.333333, 'end': 158},...

In [ ]:
model_sentence = SentenceTransformer("all-MiniLM-L6-v2")

def get_prob(prob):
  prob = prob.item()
  if prob > 0.4:
    prob = 0.99
  elif prob < -0.1:
    prob = 0.66
  else:
    prob = 0.33
  return prob

def get_soft_labels(data):
  """It returns the soft labels"""
  soft_labels = []
  aux = {'start': 10, 'prob': 0.333333, 'end': 18}
  for i in range(len(data)):
    soft_label = []
    info = data.iloc[i]
    gpt = model_sentence.encode(info['Respuesta GPT'])
    for j in info['hard_labels']:
      prob = model_sentence.similarity(model_sentence.encode(info['model_output_text'][j[0]:j[1]]), gpt)
      aux =  {'start': j[0], 'prob': get_prob(prob), 'end': j[1]}
      soft_label.append(aux)
    soft_labels.append(soft_label)
  return soft_labels


In [ ]:
merge_data['soft_labels'] = get_soft_labels(merge_data)
merge_data.head(6)

id                                        model_input  \
0  val-en-1  What did Petra van Staveren win a gold medal for?   
1  val-en-2  How many genera does the Erysiphales order con...   
2  val-en-3                   Do all arthropods have antennae?   
3  val-en-4                  When did Chance the Rapper debut?   
4  val-en-5  What is the UN Sustainable Development Goal 11...   
5  val-en-6      What are the four styles of Zhejiang cuisine?   

                                   model_output_text  \
0  Petra van Stoveren won a silver medal in the 2...   
1            The Elysiphale order contains 5 genera.   
2  Yes, all arachnids have antennas. However, not...   
3                 Chance the rapper debuted in 2011.   
4  The UN's Sustainable City initiative defines a...   
5  Zhejing cuisine is known for its unique flavor...   

                                       Respuesta GPT  \
0  Petra van Staveren won a gold medal in the 100...   
1  The Erysiphales order contains one family, Ery...   
2  Not all arthropods have antennae. While most a...   
3  Chance the Rapper debuted with his mixtape "10...   
4  The provided text does not contain information...   
5  The four styles of Zhejiang cuisine are:\n\n1....   

                                         hard_labels  \
0           [[10, 18], [25, 31], [45, 49], [69, 83]]   
1                                [[4, 14], [30, 31]]   
2             [[0, 3], [24, 32], [63, 70], [78, 87]]   
3                               [[11, 17], [29, 33]]   
4  [[21, 36], [26, 36], [55, 58], [84, 98], [103,...   
5  [[33, 39], [52, 70], [60, 70], [81, 91], [86, ...   

                                         soft_labels  
0  [{'start': 10, 'prob': 0.33, 'end': 18}, {'sta...  
1  [{'start': 4, 'prob': 0.33, 'end': 14}, {'star...  
2  [{'start': 0, 'prob': 0.33, 'end': 3}, {'start...  
3  [{'start': 11, 'prob': 0.99, 'end': 17}, {'sta...  
4  [{'start': 21, 'prob': 0.33, 'end': 36}, {'sta...  
5  [{'start': 33, 'prob': 0.33, 'end': 39}, {'sta...

In [ ]:
merge_data.iloc[0]['soft_labels']

[{'start': 10, 'prob': 0.33, 'end': 18},
 {'start': 25, 'prob': 0.33, 'end': 31},
 {'start': 45, 'prob': 0.33, 'end': 49},
 {'start': 69, 'prob': 0.33, 'end': 83}]

# Evaluation

In [ ]:
def recompute_hard_labels(soft_labels):
    """optionally, infer hard labels from the soft labels provided"""
    hard_labels = []
    prev_end = -1
    for start, end in (
        (lbl['start'], lbl['end'])
        for lbl in sorted(soft_labels, key=lambda span: (span['start'], span['end']))
        if lbl['prob'] > 0.5
    ):
        if start == prev_end:
            hard_labels[-1][-1] = end
        else:
            hard_labels.append([start, end])
        prev_end = end
    return hard_labels


def infer_soft_labels(hard_labels):
    """reformat hard labels into soft labels with prob 1"""
    return [
        {
            'start': start,
            'end': end,
            'prob': 1.0,
        }
        for start, end in hard_labels
    ]


def load_jsonl_file_to_records(filename, is_ref=True):
    """read data from a JSONL file and format that as a `pandas.DataFrame`.
    Performs minor format checks (ensures that some labels are present,
    optionally compute missing labels on the fly)."""
    df = pd.read_json(filename, lines=True)
    if not is_ref:
        assert ('hard_labels' in df.columns) or ('soft_labels' in df.columns), \
            f'File {filename} contains no predicted label!'
        if 'hard_labels' not in df.columns:
            df['hard_labels'] = df.soft_labels.apply(recompute_hard_labels)
        elif 'soft_labels' not in df.columns:
            df['soft_labels'] = df.hard_labels.apply(infer_soft_labels)
    # adding an extra column for convenience
    columns = ['id', 'soft_labels', 'hard_labels']
    if is_ref:
        df['text_len'] = df.model_output_text.apply(len)
        columns += ['text_len']
    df = df[columns]
    return df.sort_values('id').to_dict(orient='records')

def score_iou(ref_dict, pred_dict):
    """computes intersection-over-union between reference and predicted hard labels, for a single datapoint.
    inputs:
    - ref_dict: a gold reference datapoint,
    - pred_dict: a model's prediction
    returns:
    the IoU, or 1.0 if neither the reference nor the prediction contain hallucinations
    """
    # ensure the prediction is correctly matched to its reference
    assert ref_dict['id'] == pred_dict['id']
    # convert annotations to sets of indices
    ref_indices = {idx for span in ref_dict['hard_labels'] for idx in range(*span)}
    pred_indices = {idx for span in pred_dict['hard_labels'] for idx in range(*span)}
    # avoid division by zero
    if not pred_indices and not ref_indices: return 1.
    # otherwise compute & return IoU
    return len(ref_indices & pred_indices) / len(ref_indices | pred_indices)

def score_cor(ref_dict, pred_dict):
    """computes Spearman correlation between predicted and reference soft labels, for a single datapoint.
    inputs:
    - ref_dict: a gold reference datapoint,
    - pred_dict: a model's prediction
    returns:
    the Spearman correlation, or a binarized exact match (0.0 or 1.0) if the reference or prediction contains no variation
    """
    # ensure the prediction is correctly matched to its reference
    assert ref_dict['id'] == pred_dict['id']
    # convert annotations to vectors of observations
    ref_vec = [0.] * ref_dict['text_len']
    pred_vec = [0.] * ref_dict['text_len']
    for span in ref_dict['soft_labels']:
        for idx in range(span['start'], span['end']):
            ref_vec[idx] = span['prob']
    for span in pred_dict['soft_labels']:
        for idx in range(span['start'], span['end']):
            pred_vec[idx] = span['prob']
    # constant series (i.e., no hallucination) => cor is undef
    if len({round(flt, 8) for flt in pred_vec}) == 1 or len({round(flt, 8) for flt in ref_vec}) == 1 :
        return float(len({round(flt, 8) for flt in ref_vec}) == len({round(flt, 8) for flt in pred_vec}))
    # otherwise compute Spearman's rho
    return spearmanr(ref_vec, pred_vec).correlation

def main(ref_dicts, pred_dicts, output_file=None):
    assert len(ref_dicts) == len(pred_dicts)
    ious = np.array([score_iou(r, d) for r, d in zip(ref_dicts, pred_dicts)])
    cors = np.array([score_cor(r, d) for r, d in zip(ref_dicts, pred_dicts)])
    if output_file is not None:
        with open(output_file, 'w') as ostr:
            print(f'IoU: {ious.mean():.8f}', file=ostr)
            print(f'Cor: {cors.mean():.8f}', file=ostr)
    return ious, cors

In [ ]:
merge_data.to_json(f'{path}Datasets/test_ds_prueba_jsonl.jsonl', orient="records", lines = True)

In [ ]:
test_df = load_jsonl_file_to_records(f'{path}Datasets/test_ds_prueba_jsonl.jsonl')
val_df = load_jsonl_file_to_records(f'{path}Datasets/mushroom.en-val.v2.jsonl')

In [ ]:
ious, cors = main(val_df, test_df)
print(f'IoU: {ious.mean():.8f}')
print(f'Cor: {cors.mean():.8f}')

IoU: 0.30387506
Cor: 0.34864451


Con el random:

- IoU: 0.15334043
- Cor: 0.05996364

Con hard labels = palabras diferentes pero soft random:
- IoU: 0.34704504
- Cor: 0.31591253

Con hard labels = ner diferentes pero soft random:
- IoU: 0.38262133
- Cor: 0.41882317

Con hard labels = ner diferentes pero soft distancia embeddings:
- IoU: 0.30387506
- Cor: 0.34864451